test/aleph/classloader_test.clj

```diff
@@ -41,7 +40,7 @@
 (deftest test-classloader
   (testing "classloader: ensure the class loader is always a DynamicClassLoader"
     (let [result (CompletableFuture.)]
-      (with-dynamic-redefs [netty/operation-complete (partial operation-complete result)]
+      (with-redefs [netty/operation-complete (partial operation-complete result)]
```

Comment:
This is an unrelated change. Essentially, `with-dynamic-redefs` is always preferable to plain `with-redefs` in testing, and wasn't related to whether we use `bound-fn` or not.

Reply:
Not sure your comment is accurate here.
But I will have a look. With `with-dynamic-redefs`, the `netty/operation-complete` was not available on the Netty threads and it failed.

I will refactor the test a bit.

Reply:
Mhhh... I don't know exactly how to fix this.
The test was relying on the fact we were using `bound-fn*` actually.

```clojure
(let [bound-operation-complete (bound-fn* operation-complete)]   <== Here we are on a SIGINT Thread
  (.addListener f
                (reify GenericFutureListener
                  (operationComplete [_ _]
                    (ensure-dynamic-classloader)
                    (bound-operation-complete f d)))))) <== Here we are on a FastThreadLocalThread
```

Without using `bound-fn*`, I have no idea how to propagate the local binding of `operation-complete` to the `FastThreadLocalThread`. 

Any help is welcome here.

Reply:
...or did switching from `bound-fn` to `fn` in `classloader-test` fix it, because CircleCI isn't complaining?

Reply:
> Hmmm, test-classloader is failing? Are you saying it fails using with-dynamic-redefs, but works using with-redefs? 

Yes, it used to work thanks to `bound-fn*`.
This test was actually doing this job quite nicely, ensuring the bindings of the calling threads were pushed to the callback. Since we are no longer using `bound-fn*`, we won't have access to the `netty/operation-complete` rebinding which is present on the `SIGINT Thread` only.
Having to use `with-redefs` makes sense to me here.

> ...or did switching from bound-fn to fn in classloader-test fix it,

This `bound-fn` was actually to push the bindings from the `test` Thread to the `SIGINT Thread` which was then propagated with `bound-fn*` when calling operation-complete.
Since `with-redefs` is used, this is no longer necessary. (it was just a clean up commit)

> ...or did switching from bound-fn to fn in classloader-test fix it,

Tests are passing, like it was the case initially. But I have no idea how I can keep using `with-dynamic-redefs` to achieve the
same behavior...

Reply:
Well, something is wrong, because it *should* work, but I can't keep looking at it. I think it's due to the use of non-Clojure threads again (both signal and netty)

My suspicion is that `with-dynamic-redefs` is failing because internally, it relies on `binding` to store the thread-specific override map, but `binding`'s thread frames don't play well with non-Clojure threads, so I think the overrides are never seen, in which case, `with-dynamic-redefs` will fall back to the original, which won't set our test Future.

`with-redefs` works in this case because it truly replaces the root definition, instead of adding some thread-based override-enabling functionality.

Let's just go with `with-redefs` for now.

Reply:
> My suspicion is that with-dynamic-redefs is failing because internally, it relies on binding to store the thread-specific override map

That's the point. It uses local bindings. So any others threads cannot see those bindings unless we propagate.

Reply:
That's only true for non-Clojure threads. And there's no requirement for that, it should be possible to place the override machinery directly in the root var, so it would work, regardless of thread.

Reply:
Hmm, `with-dynamic-redefs` is more limited than I recall. It only works with fns, for starters. Might be able to use `bound-fn` tho.

Real shame Hickey started requiring `^:dynamic` back in 1.3. Used to be simpler.